File: Database/analysis.js
const apicomm = require("./adapters/apihandler");
const dbhandler = require("./adapters/dbhandler.js");
const scrapping = require("./adapters/scrapinghandler");
const config = require("./config.json");

function abbreviateName(name) {
  if (name.includes(".")) {
    return name;
  }
  return name
    .split(" ")
    .map((p) => `${p[0]}.`)
    .join(" ");
}

async function run() {
  const season = config.currentSeason;

  const db = new dbhandler.Database();
  await db.init();

  const response = await apicomm.nhlApiRequest(
    `/api/v1/standings?season=${season}`
  );

  try {
    const teamAnalysisCollection = db.getCollection("analysis");

    for (let record of response.records) {
      for (let teamRecord of record.teamRecords) {
        const playerStats = await apicomm.nhlApiRequest(
          `/api/v1/teams/${teamRecord.team.id}?hydrate=roster(season=${season},person(stats(splits=statsSingleSeason)))`
        );

        const skaterScoringStats = await apicomm.enhancedNhlApiRequest(
          `/stats/rest/en/skater/scoringRates?isAggregate=false&isGame=false&start=0&limit=50&factCayenneExp=gamesPlayed>=1&cayenneExp=teamId=${teamRecord.team.id} and gameTypeId=2 and seasonId<=${season} and seasonId>=${season}`
        );

        const skaterPowerplayStats = await apicomm.enhancedNhlApiRequest(
          `/stats/rest/en/skater/powerplay?isAggregate=false&isGame=false&start=0&limit=50&factCayenneExp=gamesPlayed>=1&cayenneExp=teamId=${teamRecord.team.id} and gameTypeId=2 and seasonId<=${season} and seasonId>=${season}`
        );

        const enhancedGoalieStats = await apicomm.enhancedNhlApiRequest(
          `/stats/rest/en/goalie/advanced?isAggregate=false&isGame=false&start=0&limit=50&factCayenneExp=gamesPlayed>=1&cayenneExp=teamId=${teamRecord.team.id} and gameTypeId=2 and seasonId<=${season} and seasonId>=${season}`
        );

        const enhancedSkaterStats = [
          ...skaterScoringStats.data
            .concat(skaterPowerplayStats.data)
            .reduce(
              (m, o) =>
                m.set(o.playerId, Object.assign(m.get(o.playerId) || {}, o)),
              new Map()
            )
            .values(),
        ];

        const playersRoster = playerStats.teams[0].roster.roster;
        const fmtRoster = playersRoster
          .filter((p) => {
            return (
              p.person.stats[0].splits != null &&
              p.person.stats[0].splits.length > 0
            );
          })
          .map((p) => {
            let enhancedStats = enhancedSkaterStats.find(
              (enhancedStats) => enhancedStats.playerId === p.person.id
            );
            if (!enhancedStats) {
              enhancedStats = enhancedGoalieStats.data.find(
                (enhancedStats) => enhancedStats.playerId === p.person.id
              );
            }
            return {
              id: p.person.id,
              abbrName: `${abbreviateName(p.person.firstName)} ${
                p.person.lastName
              }`,
              fullName: p.person.fullName,
              currentAge: p.person.currentAge,
              stats: p.person.stats[0].splits[0].stat,
              advancedStats: enhancedStats,
            };
          });

        const lines = await scrapping.scrapLines(teamRecord.team.name);

        const teamStats = await apicomm.nhlApiRequest(
          `/api/v1/teams/${teamRecord.team.id}/stats`
        );
        const options = { upsert: true };
        const filter = { id: teamRecord.team.id };
        const rankings = teamStats.stats[1].splits[0].stat;
        Object.keys(rankings).forEach(function (key) {
          rankings[key] = parseInt(rankings[key]);
        });

        const updateDoc = {
          $set: {
            id: teamRecord.team.id,
            team: teamRecord.team,
            leagueRecord: teamRecord.leagueRecord,
            points: teamRecord.points,
            regulationWins: teamRecord.regulationWins,
            leagueRank: parseInt(teamRecord.leagueRank),
            leagueL10Rank: parseInt(teamRecord.leagueL10Rank),
            leagueHomeRank: parseInt(teamRecord.leagueHomeRank),
            leagueRoadRank: parseInt(teamRecord.leagueRoadRank),
            ppLeagueRank: parseInt(teamRecord.ppLeagueRank),
            divisionRank: parseInt(teamRecord.divisionRank),
            divisionL10Rank: parseInt(teamRecord.divisionL10Rank),
            divisionHomeRank: parseInt(teamRecord.divisionHomeRank),
            divisionRoadRank: parseInt(teamRecord.divisionRoadRank),
            ppDivisionRank: parseInt(teamRecord.ppDivisionRank),
            conferenceRank: parseInt(teamRecord.conferenceRank),
            conferenceL10Rank: parseInt(teamRecord.conferenceL10Rank),
            conferenceHomeRank: parseInt(teamRecord.conferenceHomeRank),
            conferenceRoadRank: parseInt(teamRecord.conferenceRoadRank),
            ppConferenceRank: parseInt(teamRecord.ppConferenceRank),
            row: teamRecord.row,
            statsSingleSeason: teamStats.stats[0].splits[0].stat,
            regularSeasonStatRankings: rankings,
            rosterStats: fmtRoster,
            lines,
            lastUpdated: teamRecord.lastUpdated,
          },
        };
        try {
          const queryResult = await teamAnalysisCollection.updateOne(
            filter,
            updateDoc,
            options
          );
          console.log(
            `${queryResult.matchedCount} document(s) matched the filter, updated ${queryResult.modifiedCount} document(s): ${teamRecord.team.name}`
          );
        } catch (ex) {
          console.log(ex);
        }
      }
    }
  } finally {
    await db.closeClient();
  }
}

run().catch(console.dir);

File: Database/analysis.py
from datetime import datetime
from apihandler import nhl_api_request, enhanced_nhl_api_request
from dbhandler import get_collection
from collections import defaultdict
from utils import get_prop
from constants import DEFAULT_SEASON
import re

def abbrev_name(first_name, last_name):
  return "{0}. {1}".format(first_name[0], last_name) 

def parse_ordinals(stats):
  dictionary = {}
  for prop, value in stats.items():
    dictionary[prop] = int(re.sub("[^0-9]", "", value))
  return dictionary

collection = get_collection("analysis")

season = DEFAULT_SEASON
standings = nhl_api_request("/api/v1/standings", {"season": season})
records = standings["records"]



for record in records:
  for teamRecord in record["teamRecords"]:
    player_stats = nhl_api_request("/api/v1/teams/{0}".format(teamRecord["team"]["id"]), { "hydrate": "roster(season={0},person(stats(splits=statsSingleSeason)))".format(season)})

    enhanced_query_options =  {"isAggregate": "false", "isGame": "false", "start": "0", "limit": "50", "factCayenneExp": "gamesPlayed>=1", "cayenneExp": "teamId={0} and gameTypeId=2 and seasonId <= {1} and seasonId >= {1}".format(teamRecord["team"]["id"], season)}
    skater_scoring_stats = enhanced_nhl_api_request("/stats/rest/en/skater/scoringRates", enhanced_query_options)["data"]
    skater_pp_stats =  enhanced_nhl_api_request("/stats/rest/en/skater/powerplay", enhanced_query_options)["data"]
    skater_pk_stats = enhanced_nhl_api_request("/stats/rest/en/skater/penaltykill", enhanced_query_options)["data"]
    skater_sat_stats = enhanced_nhl_api_request("/stats/rest/en/skater/summaryshooting", enhanced_query_options)["data"]
    skater_puck_stats = enhanced_nhl_api_request("/stats/rest/en/skater/puckPossessions", enhanced_query_options)["data"]
    enhanced_goalies_stats =  enhanced_nhl_api_request("/stats/rest/en/goalie/advanced", enhanced_query_options)["data"]
    
    enhanced_skaters = defaultdict(list)
    enhanced_goalies = defaultdict(list)
    for skater_scoring in skater_scoring_stats:
      pp_stats = next((x for x in skater_pp_stats if x["playerId"] == skater_scoring["playerId"]), None)
      pk_stats = next((x for x in skater_pk_stats if x["playerId"] == skater_scoring["playerId"]), None)
      sat_stats = next((x for x in skater_sat_stats if x["playerId"] == skater_scoring["playerId"]), None)
      puck_stats = next((x for x in skater_puck_stats if x["playerId"] == skater_scoring["playerId"]), None)
      enhanced_skaters[skater_scoring["playerId"]] = skater_scoring | pp_stats | pk_stats | sat_stats | puck_stats
    for goalie_advanced in enhanced_goalies_stats:
      enhanced_goalies[goalie_advanced["playerId"]] = goalie_advanced
    # filter players that didn't play at all
    players_roster = player_stats["teams"][0]["roster"]["roster"]
    players_playing = [a for a in players_roster if len(a["person"]["stats"][0]["splits"]) > 0]
    players_playing_enhanced_stats = []
    for playing_player in players_playing:
      player_stats = None
      playing_player_id = playing_player["person"]["id"]
      for skater_id, skater_stats in enhanced_skaters.items():
        if skater_id == playing_player_id:
          player_stats = skater_stats
          break
      if (player_stats is None):
        for goalie_id, goalie_stats in enhanced_goalies.items():
          if goalie_id == playing_player_id:
            player_stats = goalie_stats
            break
      person = playing_player["person"]
      players_playing_enhanced_stats.append({
        "id": person["id"],
        "abbrName": abbrev_name(person["firstName"], person["lastName"]) ,
        "fullName": person["fullName"],
        "primaryNumber": get_prop(person,"primaryNumber"),
        "positionName":person["primaryPosition"]["name"],
        "nationality": person["nationality"],
        "currentAge": get_prop(person,"currentAge"),
        "stats": person["stats"][0]["splits"][0]["stat"],
        "advancedStats": player_stats
      })

    team_stats = nhl_api_request("/api/v1/teams/{0}/stats".format(teamRecord["team"]["id"]), {})
    filter = { "id": teamRecord["team"]["id"] }
    
    try:
      update = {
        "$set": {
          "id": teamRecord["team"]["id"],
          "team": teamRecord["team"],
          "leagueRecord": teamRecord["leagueRecord"],
          "points": teamRecord["points"],
          "regulationWins": teamRecord["regulationWins"],
          "leagueRank": int(teamRecord["leagueRank"]),
          "leagueL10Rank": int(teamRecord["leagueL10Rank"]),
          "leagueHomeRank": int(teamRecord["leagueHomeRank"]),
          "leagueRoadRank": int(teamRecord["leagueRoadRank"]),
          "ppLeagueRank": int(teamRecord["ppLeagueRank"]),
          "divisionRank": int(teamRecord["divisionRank"]),
          "divisionL10Rank": int(teamRecord["divisionL10Rank"]),
          "divisionHomeRank": int(teamRecord["divisionHomeRank"]),
          "divisionRoadRank": int(teamRecord["divisionRoadRank"]),
          "ppDivisionRank": int(teamRecord["ppDivisionRank"]),
          "conferenceRank": int(teamRecord["conferenceRank"]),
          "conferenceL10Rank": int(teamRecord["conferenceL10Rank"]),
          "conferenceHomeRank": int(teamRecord["conferenceHomeRank"]),
          "conferenceRoadRank": int(teamRecord["conferenceRoadRank"]),
          "ppConferenceRank": int(teamRecord["ppConferenceRank"]),
          "row": teamRecord["row"],
          "statsSingleSeason": team_stats["stats"][0]["splits"][0]["stat"],
          "regularSeasonStatRankings": parse_ordinals(team_stats["stats"][1]["splits"][0]["stat"]),
          "rosterStats": players_playing_enhanced_stats,
          "lastUpdate": datetime.now()
        }
      }
      result = collection.update_one(filter, update, True)
      print("updated analysis for: {0}".format(teamRecord["team"]["name"]))
    except Exception as e:
      print("exception occured")
      print(e)
